### PR TITLE
Implement BeanInitializer

### PR DESCRIPTION
**Pull request by NazarSmith228**
## Desctiprion:
`BeanInitializer` is a class that is responsible for triggering bean initialization. It does that by fetching all bean dependencies from the bean definition container and passing those into `BeanDefinition.instance()` method.
### Changes:

1. Implement BeanInitializer, based on Google Guava Table. 
2. Add a code style folder, where the latest code scheme .xml file is stored.

## Ticket: https://trello.com/c/Z96MzpAi

### CODE REVIEW

**Comment by DanielChanger** [comment]
Please, add PR description and ticket link as I did in my PRs

File: src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java [hunk 1]
```diff
@@ -7,7 +7,13 @@ public class BeanInitializer {
     /**
      * Instantiates beans and injects dependencies in them
      */
-    void initialize(Map<String, BeanDefinition> beanDefinitions) {
-
+    public void initialize(Map<String, BeanDefinition> beanDefinitions) {
+        for (Map.Entry<String, BeanDefinition> entry : beanDefinitions.entrySet()) {
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java]: Looks like a perfect candidate for 'var' instead of 'Map.Entry<String, BeanDefinition> '

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static org.mockito.Mockito.*;
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Let's avoid usage of * to be more explicit

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+                        Map<BeanDefinition, BeanDefinition[]> expectedInvocationArgs, String description) {
+        var beanInitializer = new BeanInitializer();
+        beanInitializer.initialize(beanDefinitionMap);
+
+        beanDefinitionMap.values().forEach(bd -> verify(bd).instance(expectedInvocationArgs.get(bd)));
+    }
+
+
+    private static BeanDefinition prepare(String name, List<String> dependenciesNames) {
+        BeanDefinition beanDefinition = spy(BeanDefinition.class);
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Why are we using spy here? Why not to use a simple mock?

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+                        ),
+                        "Bean initialization default test case"
+                )
+        );
+    }
+
+    @ParameterizedTest(name = "[{index}] - {2}")
+    @MethodSource("testArgs")
+    void testInitialize(Map<String, BeanDefinition> beanDefinitionMap,
+                        Map<BeanDefinition, BeanDefinition[]> expectedInvocationArgs, String description) {
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Can we remove this description parameter, since it's not used? It worked for me at least

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    private static final String BD3 = "beanDef3";
+    private static final String BD4 = "beanDef4";
+
+    private static Stream<Arguments> testArgs() {
+        BeanDefinition beanDef1 = prepare(BD1, List.of(BD2, BD3));
+        BeanDefinition beanDef2 = prepare(BD2, List.of(BD3, BD4));
+        BeanDefinition beanDef3 = prepare(BD3, singletonList(BD4));
+        BeanDefinition beanDef4 = prepare(BD4, emptyList());
+
+        return Stream.of(
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: One suggestion:  don't one-line all these args in one Arguments.of() method. It would be more readable if you assign each of the args to variables and name them accordingly

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    @ParameterizedTest(name = "[{index}] - {2}")
+    @MethodSource("testArgs")
+    void testInitialize(Map<String, BeanDefinition> beanDefinitionMap,
+                        Map<BeanDefinition, BeanDefinition[]> expectedInvocationArgs, String description) {
+        var beanInitializer = new BeanInitializer();
+        beanInitializer.initialize(beanDefinitionMap);
+
+        beanDefinitionMap.values().forEach(bd -> verify(bd).instance(expectedInvocationArgs.get(bd)));
```
Comment (DanielChanger) [inline src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Can we use new lines on method chaining? Could you please update code style xml for this and use it in this PR?

File: src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java [hunk 7]
```diff
@@ -7,7 +7,13 @@ public class BeanInitializer {
...
+            BeanDefinition beanDefinition = entry.getValue();
+            BeanDefinition[] instanceDependencies = beanDefinition.dependenciesNames().stream()
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java]: Can we align method chains like this below, and also update code style xml and push it in this PR?
```
beanDefinition.dependenciesNames()
              .stream()
              .map(beanDefinitions::get)
              .toArray(BeanDefinition[]::new);
```

File: pom.xml [hunk 8]
```diff
@@ -9,13 +9,14 @@
     <version>1.0-SNAPSHOT</version>
 
     <properties>
-        <maven.compiler.source>17</maven.compiler.source>
-        <maven.compiler.target>17</maven.compiler.target>
+        <java.version>17</java.version>
+        <maven-compiler-plugin.version>3.10.1</maven-compiler-plugin.version>
```
Comment (DanielChanger) [inline pom.xml]: Could you explain why do we need it, please? You just want to move those properties to the plugin configuration? If that's the case I'm good with the change

**Review by DanielChanger — CHANGES_REQUESTED** [review]
Looks good to me overall. Only minor issues, which still need to be addressed.

**Comment by DanielChanger** [comment]
For now you can ignore sonar, I'm going to handle this.

**Comment by DanielChanger** [comment]
Also, don't forget to notify the team each time you create a PR

**Comment by DanielChanger** [comment]
BTW, let's not duplicate ticket number in PR title, just for consistency across PRs

**Comment by DanielChanger** [comment]
Don't forget to leave logs

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+import org.junit.jupiter.params.provider.MethodSource;
+
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+import static java.util.Collections.emptyList;
+import static java.util.Collections.singletonList;
+import static org.mockito.Mockito.*;
+
```
Comment (nourish1709) [inline src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java:19]: I thought we'd need `@ExtendWith(MockitoExtension.class)` to make it work and be able to use mocks. How does it work without this extension?

Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java:19]: It's required if you want to use `@Mock` annotation, otherwise you can omit it, like I did in my PR

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 4 above)
Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: @DanielChanger it is used in the test name, not in the code.

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 2 above)
Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: @DanielChanger I doubt it makes sense. Do you think `mock()` is not explicit enough? Mockito is the first framework you think of, when it comes to mocks, spies etc. Apart from that, I don't think that it is worth checking whether each and every PR doesn`t have star imports

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 5 above)
Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: @DanielChanger wrong suggestion :) Each entry of `Arguments.of()` should contain input data and expected results ( bean definition map and expected invocation arguments in our case). Since there are no corner cases, out test has only one entry of arguments.

File: pom.xml (hunk 8 above)
Comment (NazarSmith228) [reply pom.xml]: @DanielChanger exactly. I've added them in order to use `maven-compiler-plugin`

File: src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java (hunk 7 above)
Comment (NazarSmith228) [reply src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java]: @DanielChanger yup, fixed

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 2 above)
Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: I say that you should use static imports without * 
`import static org.mockito.Mockito.mock;`

I'm not saying you have to use `Mockito.mock()` in your code

You'll still be able to use just `mock()` without class

By removing * we're just adding explicitness to imports, it's considered to be a good practice

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 6 above)
Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: @DanielChanger well, since there are only 2 methods in the chain, and they perfectly fit into one line, I don't think that new line will improve readability: 
image
Do you think it will?

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 4 above)
Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Okay, I noticed it really affects test name during execution, so it's fine here

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 3 above)
Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: @DanielChanger I forgot to remove it) In fact, `mock()` is enough in this case

File: src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java (hunk 1 above)
Comment (NazarSmith228) [reply src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java]: @DanielChanger yup, makes sense

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 5 above)
Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: No, I mean just assign first map to a variable, and second one, to give them kind of names, so it's immediately clear for everybody reading the code where's expected and input data.

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 3 above)
Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Exactly

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 6 above)
Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Okay I agree, but aligning methods is what we'd really like to have (not specifically here though, if you don't use new line).

What I'd suggest here is moving expectedInvocationArgs.get(bd) to a variable called dependencies or whatever you choose.
It would simplify 52 line for sure

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 2 above)
Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: What if we have 5+ static imports of the same class? Or 10+? Should we also avoid star imports?

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 6 above)
Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: image
Still don't think it is a good idea :) I prefer one-line lambdas, but I can rewrite it like on screenshot if you wish)

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 5 above)
Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Names are visible in test method parameters :) `Arguments` method name is used in `@MethodSource` annotation. Thus, it is quite obvious, what each parameter is used for. If such approach is not clear for everybody, then I would recommend to avoid the use of `@ParameterizedTest`

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 2 above)
Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: it depends. In my practice only Sonar could prevent from a big amount of imports. 
But often it just a sign that you have to split your class. 
So, in this case I see no reason not to follow commonly accepted good practice.

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 6 above)
Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Yeah, I didn't notice you call get with bd param. Thus it's not possible

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 5 above)
Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: When you read code statically you don't want to jump here and there to know what arguments are. Moreover, people who are not aware of parametrized feature maybe won't even know where to jump for some period of time :) Like, of course they'll eventually find those parameters, but wouldn't it be better to cause immediate reaction of understanding than confusing everybody who is not familiar with parametrized test?

Moreover, one other thing is that Arguments.of() call is scattered 10 line while it, so it looks enormous

```
Arguments.of(
                        Map.of(
                                BD1, beanDef1, BD2, beanDef2,
                                BD3, beanDef3, BD4, beanDef4
                        ),
                        Map.of(
                                beanDef1, new BeanDefinition[]{beanDef2, beanDef3},
                                beanDef2, new BeanDefinition[]{beanDef3, beanDef4},
                                beanDef3, new BeanDefinition[]{beanDef4},
                                beanDef4, new BeanDefinition[0]
                        ),
                        "Bean initialization default test case"
                )
```

**Okay, I start thinking that it can be a very long discussion, which is based more on personal preferences than on the real business here. So, if you're comfortable with it, let it be as it is.**

**Comment by DanielChanger** [comment]
Please, rebase on main, so that your build become green

File: src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java (hunk 5 above)
Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: This is just a regular  `@ParameterizedTest` usage. It is always written like that. Our conversation has shown, that it is a bad idea to use this JUnit feature in our project. Therefore, I will rewrite this test using regular approach

Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: No....

Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: Maybe I wasn't that clear, because we're not on the same page now. Should we have a call?

Comment (NazarSmith228) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java]: @DanielChanger let’s do it tomorrow after our lesson. I don’t wanna miss it because of lack of sleep 😁

Comment (DanielChanger) [reply src/test/java/com/bobocode/hoverla/bring/context/BeanInitializerTest.java:28]: It's 5 minutes

File: pom.xml [hunk 10]
```diff
@@ -54,6 +60,7 @@
             <groupId>org.assertj</groupId>
             <artifactId>assertj-core</artifactId>
             <version>${assertj.version}</version>
+            <scope>test</scope>
```
Comment (DanielChanger) [inline pom.xml:63]: Good catch

File: src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java [hunk 11]
```diff
@@ -1,13 +1,40 @@
...
 
-import java.util.Map;
+import com.google.common.collect.Table;
+import lombok.extern.slf4j.Slf4j;
 
+import java.util.Collection;
+import java.util.Objects;
+
+/**
+ * Class, responsible for triggering bean initialization.
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java:9]: (nit) comma can be omitted

File: src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java [hunk 12]
```diff
@@ -1,13 +1,40 @@
...
+        log.debug("Bean initialization started");
+        Collection<BeanDefinition> beanDefinitions = beanDefinitionsTable.values();
+
+        for (BeanDefinition beanDefinition : beanDefinitions) {
+            log.trace("Resolving dependencies for bean definition: {} - {}", beanDefinition.name(), beanDefinition.type());
+            BeanDefinition[] beanDependencies = beanDefinition.dependencies()
+                    .entrySet()
+                    .stream()
+                    .map(keyPair -> beanDefinitionsTable.get(keyPair.getKey(), keyPair.getValue()))
+                    .filter(Objects::nonNull)
```
Comment (DanielChanger) [inline src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java]: (nit) seems like we don't need this check, because dependency missing will be handled in validator

**Review by DanielChanger — DISMISSED** [review]

File: src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java (hunk 12 above)
Comment (NazarSmith228) [reply src/main/java/com/bobocode/hoverla/bring/context/BeanInitializer.java]: @DanielChanger double check to be sure, nobody wants NPE)

**Review by DanielChanger — APPROVED** [review]

**Review by nadya-gainutsa — APPROVED** [review]
